src/traefik.py

```diff
@@ -0,0 +1,574 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Traefik workload interface."""
+import contextlib
+import enum
+import logging
+import re
+import socket
+from pathlib import Path
+from string import Template
+from typing import Any, Dict, List, Optional, cast
+
+import yaml
+from lightkube.models.core_v1 import ServicePort
+from ops import Container
+from ops.pebble import LayerDict, PathError
+from utils import is_hostname
+
+logger = logging.getLogger(__name__)
+DYNAMIC_CONFIG_DIR = "/opt/traefik/juju"
+STATIC_CONFIG_DIR = "/etc/traefik"
+STATIC_CONFIG_PATH = f"{STATIC_CONFIG_DIR}/traefik.yaml"
+DYNAMIC_CERTS_PATH = f"{DYNAMIC_CONFIG_DIR}/certificates.yaml"
+DYNAMIC_TRACING_PATH = f"{DYNAMIC_CONFIG_DIR}/tracing.yaml"
+SERVER_CERT_PATH = f"{DYNAMIC_CONFIG_DIR}/server.cert"
+SERVER_KEY_PATH = f"{DYNAMIC_CONFIG_DIR}/server.key"
+CA_CERTS_PATH = "/usr/local/share/ca-certificates"
+CA_CERT_PATH = f"{CA_CERTS_PATH}/traefik-ca.crt"
+RECV_CA_TEMPLATE = Template(f"{CA_CERTS_PATH}/receive-ca-cert-$rel_id-ca.crt")
+BIN_PATH = "/usr/bin/traefik"
+
+
+class RoutingMode(enum.Enum):
+    """Routing mode."""
+
+    path = "path"
+    subdomain = "subdomain"
+
+
+class TraefikError(RuntimeError):
+    """Base class for errors raised by this module."""
+
+
+class ContainerNotReadyError(TraefikError):
+    """Raised if the caller incorrectly assumes that the traefik container is ready."""
+
+
+class Traefik:
+    """Traefik workload representation."""
+
+    diagnostics_port = 8082  # Prometheus metrics, healthcheck/ping
+    _port = 80
+    _tls_port = 443
+    log_path = "/var/log/traefik.log"
```

Comment:
Now that we have a dedicated module, these vars that are not pertaining to the class itself could become module scope vars (with `_` prefix).
`_layer_name` I'd still keep here.

Not critical.